src/Compilers/CSharp/Portable/Lowering/StateMachineRewriter/SynthesizedStateMachineMethod.cs

```diff
@@ -9,19 +10,18 @@ namespace Microsoft.CodeAnalysis.CSharp
     /// <summary>
     /// State machine interface method implementation.
     /// </summary>
-    internal sealed class SynthesizedStateMachineMethod : SynthesizedImplementationMethod, ISynthesizedMethodBodyImplementationSymbol
+    internal abstract class SynthesizedStateMachineMethodBase : SynthesizedImplementationMethod, ISynthesizedMethodBodyImplementationSymbol
```

Comment:
I suggest `SynthesizedStateMachineMethod` for the common (abstract) base, `SynthesizedStateMachineMoveNextMethod` and `SynthesizedStateMachineDebuggerHiddenMethod` for the two concrete classes. In that case you could either continue to test the .Name property, or add a virtual property `IsMoveNext` for the places where you need that test.

Alternately, if you're really ambitious, you could move some of the dependent behavior into virtual methods, but that seems far out of scope for this PR.